### ISSUE TITLE
Feature request: color-coded graphs for performance visualization

### ISSUE THREAD

**Issue by legel** [johnmarktaylor91/torchlens#20]
Gathering data from 
https://pytorch.org/tutorials/recipes/recipes/profiler_recipe.html

...it would be fantastic if there was a library with a one-line API comparable to what Tensorboard previously offered with TensorFlow, for color-coded graph visualization of performance metrics per computational graph element -- namely, runtime, but also of interest, would be memory metrics... e.g. see https://branyang.gitbooks.io/tfdocs/content/get_started/graph_viz.html

The problem with Tensorboard PyTorch support is apparently it's a mess right now...
Please ping me if this is of interest to develop, I think it would greatly help ML developers to be able to both visualize graphs and visualize performance bottlenecks of the graphs...

**Comment by legel** [comment]
PS probably it's obvious, by color-coding, I mean, e.g. 

Gradient from Blue to Red, where the darkest blue == max seconds of processing time, darkest red == least seconds, based on a simple min/max normalization from all computed graph elements, and a best-estimate allocation of the Profile runtimes per element, shown on your graph viz...

**Comment by johnmarktaylor91** [comment]
This is a great suggestion, thanks so much! TorchLens already logs all that info, so it’s just a matter of allowing the visuals to reflect it. Currently the color scheme is one that I devised to try to make the salient aspects of the network pop out, with minimal tinkering from the user. But, it could be a helpful “power feature” to allow the user to override these defaults as needed.

The tricky thing would be balancing simplicity and flexibility. Currently, my philosophy is to be conservative about hard-coding new use-cases, while giving users all the data and flexibility they need to do anything they want. 

What about something like the following: provide a set of optional visualization arguments for variables like the color, shape, and size of the box for each layer. For each such argument, allow the user to pass in a function for determining the value of that variable. For instance,

`def color_func(model_history, layer):
         rgb = some_func(layer.time_elapsed)
         return rgb

show_model_graph(model, x, color_func=color_func)’

So, the user would pass in a function that takes as input the data structure both for the overall model, and for each individual layer, which would compute the visualization variables based on the metadata provided by TorchLens.

I’ll have to think about this, but does this seem sufficient for your use case?

**Comment by legel** [comment]
That makes sense!

I would instantly use that.

When I had the vision of color-coding a graphical network representing Torch runtime / memory stats, etc., I also posted it to a related open source project called TorchExplorer.  There I also published a [visual demo](https://github.com/spfrommer/torchexplorer/issues/51#issuecomment-1880198763) based on a Turbo colormap, including Python code, which simulates a dummy neural network I put together for visual purposes.  I'd love to port that same exact code to real graphical networks of very complicated neural networks I'm trying to profile now (LLMs, vision transformers, ...)

Let me know if you push any code for custom color-coding based on e.g runtime, I'd love to follow-up and share some examples with fine-tuned colormaps that I think could look really awesome (like below) and be very useful.

![neural_network_color_coded](https://github.com/johnmarktaylor91/torchlens/assets/1915466/5e88ecd3-5c8a-4b45-9633-ce6fef3cb921)

**Comment by legel** [comment]
Just showing the high-level graph and big ass tables you get from TensorBoard for runtime and memory visualization of an exported PyTorch Profiler of CUDA... almost useless for gathering broader network understanding in the context of what modules are consuming most memory and why...

<img width="1481" alt="Screen Shot 2024-01-09 at 12 16 40 AM" src="https://github.com/johnmarktaylor91/torchlens/assets/1915466/9cbce55c-4318-4cac-a1c5-0253f8134db4">

<img width="1482" alt="Screen Shot 2024-01-09 at 12 15 41 AM" src="https://github.com/johnmarktaylor91/torchlens/assets/1915466/e8aa7dbb-e71e-4f17-9130-cf0f3425403a">

<img width="1292" alt="Screen Shot 2024-01-09 at 12 14 49 AM" src="https://github.com/johnmarktaylor91/torchlens/assets/1915466/ff0c876d-6cb5-48d2-b378-178b23557727">

**Comment by johnmarktaylor91** [comment]
I'll put it on top of my list and keep you posted once it's implemented! Can't wait to see what you do with it.

**Comment by legel** [comment]
Sounds great!
Yeah I’ll be happy to share code and demo(s) with the core infra you setup for coloring the compute graph.

**Comment by johnmarktaylor91** [comment]
Okay, here's the interface after some initial tinkering... does this seem natural and intuitive to your eyes? Basically, you provide a dict specifying the visualization arguments you want to override (these correspond to the various graphviz functions), where each value is either a literal value (for things that don't depend on the data in the model or layer) or a function (for things that depend on metadata about the model or layer). 

`model = torchvision.models.alexnet()
x = torch.randn(1, 3, 224, 224)

def colorize_runtime(model_history, layer):
    cmap = matplotlib.cm.get_cmap('jet')
    max_runtime = max([layer.func_time_elapsed for layer in model_history])
    min_runtime = min([layer.func_time_elapsed for layer in model_history])
    scale_pos = (layer.func_time_elapsed - min_runtime) / (max_runtime - min_runtime)
    hex_color = matplotlib.colors.to_hex(cmap(scale_pos))
    return hex_color

def label_layer(model_history, layer):
    return layer.layer_label

# Color each node based on runtime, remove all text except layer label.
# Logic: if the provided value is a function, call it on the model history and layer. Otherwise, just use the value.
vis_node_overrides = {'label': label_layer,
                      'shape': 'ellipse',
                      'fillcolor': colorize_runtime}

# Remove graph label
vis_graph_overrides = {'label': 'Runtime-colorized visualization'}

# Hide the nested module boxes by turning them white and removing the text.
vis_module_overrides = {'pencolor': 'white',
                        'label': ''}

tl.show_model_graph(model, x, vis_node_overrides=vis_node_overrides, vis_graph_overrides=vis_graph_overrides,
                    vis_module_overrides=vis_module_overrides)`

![image](https://github.com/johnmarktaylor91/torchlens/assets/16569815/30821820-8759-454f-95b3-0616301383cf)

**Comment by johnmarktaylor91** [comment]
Here's another example where the nodes are colored based on their runtime, and sized based on their filesize:

![image](https://github.com/johnmarktaylor91/torchlens/assets/16569815/f4343074-f534-4731-8757-3a6f90496b51)

**Comment by legel** [comment]
These both look great!

In terms of the API you propose, I think it's simple and practical enough, so long as the intent of being able to do custom color-coding (and size-coding -- awesome work) is clear.

You might consider wrapping up these examples with your final API in a demo Jupyter notebook, with a `notebooks` section at the top-level, e.g. as Meta DinoV2 researchers did here: https://github.com/facebookresearch/dinov2/tree/main/notebooks

I think `turbo` cmap in Matplotlib should be superior but very similar to `jet`, and you can just change text to that.

The size-coding is super nice and innovative.  Most intuitively, it seems like size can do a great job of showing off differences in _memory_ requirements.  I think that's what you've started with in terms of `layer.tensor_fsize` but am not familiar enough with that API to know if `tensor_fsize` corresponds to actual GPU VRAM or RAM consumption or what.  (Ideally, it could be useful to specify "CUDA" or "CPU" memory consumption per layer, if that's indeed what you're going for.)

In general, I think a really good legend will be key to bringing together the colors and sizes into a useful data visualization exploration.  That's because it should be possible to quickly reference the colors/size vs. a ground-truth legend.  So it will be great to have a big legend that shows both a range of colors across full colormap (as in my example, above) as well as a similar legend custom made for showing the min/max sizes with numerical values alongside.

This is practically already very useful, and I'm excited to pick it up and share some demos with some of the complex neural net modules I'm currently seeking to optimize performance of...

**Comment by legel** [comment]
PS I just saw that you're a postgrad researcher at Columbia.  I studied ML there for comp sci grad school :)

**Comment by johnmarktaylor91** [comment]
Apologies for the slow rollout on this, but this functionality has now been added to the main branch in the latest release.

**Comment by legel** [comment]
Cool! 
I'll make sure to check this out when I'm hitting my next Torch performance bottleneck. :) 
Cheers!

**Comment by johnmarktaylor91** [comment]
I welcome any feedback you've got! Closing for now, cheers
